File: integration_tests/dart_client_tests/test/lists_test.dart
import 'package:test/test.dart';
import 'package:graphlink_client_integration_tests/generated/client/graph_link_client.dart';
import 'package:graphlink_client_integration_tests/generated/enums/user_status.dart';
import 'fixtures.dart';
import 'mock_adapter.dart';

void main() {
  late MockAdapter adapter;
  late GraphLinkClient client;

  setUp(() {
    adapter = MockAdapter();
    client = GraphLinkClient(adapter: adapter.call, wsAdapter: MockWebSocketAdapter());
  });

  group('non-nullable list of objects', () {
    test('listUsers returns a correctly typed list', () async {
      adapter.registerData('listUsers', {
        'listUsers': [kUserAliceJson, kUserBobJson],
      });
      final res = await client.queries.listUsers();
      expect(res.listUsers.length, equals(2));
      expect(res.listUsers[0].id, equals('user-1'));
      expect(res.listUsers[1].id, equals('user-2'));
    });

    test('empty list is returned as empty list, not null', () async {
      adapter.registerData('listUsers', {'listUsers': []});
      final res = await client.queries.listUsers();
      expect(res.listUsers, isNotNull);
      expect(res.listUsers, isEmpty);
    });

    test('each item in the list is fully deserialized', () async {
      adapter.registerData('listUsers', {
        'listUsers': [kUserAliceJson],
      });
      final res = await client.queries.listUsers();
      final user = res.listUsers.first;
      expect(user.name, equals('Alice Smith'));
      expect(user.address.city, equals('Springfield'));
    });
  });

  group('list of scalars on a type', () {
    test('User.tags list deserializes correctly', () async {
      adapter.registerData('getUser', {'getUser': kUserAliceJson});
      final res = await client.queries.getUser(id: 'user-1');
      expect(res.getUser.tags, equals(['admin', 'beta']));
    });

    test('empty tags list is empty, not null', () async {
      adapter.registerData('getUser', {'getUser': kUserBobJson});
      final res = await client.queries.getUser(id: 'user-2');
      expect(res.getUser.tags, isEmpty);
    });

    test('nullable scores list is null when server returns null', () async {
      adapter.registerData('getUser', {'getUser': kUserBobJson});
      final res = await client.queries.getUser(id: 'user-2');
      expect(res.getUser.scores, isNull);
    });

    test('nullable scores list has values when server returns them', () async {
      adapter.registerData('getUser', {'getUser': kUserAliceJson});
      final res = await client.queries.getUser(id: 'user-1');
      expect(res.getUser.scores, equals([10, 20, 30]));
    });
  });

  group('list of objects with nullable fields', () {
    test('getTags returns tags with nullable color', () async {
      adapter.registerData('getTags', {'getTags': kTagsJson});
      final res = await client.queries.getTags();
      expect(res.getTags.length, equals(2));
      expect(res.getTags[0].label, equals('dart'));
      expect(res.getTags[0].color, equals('#0175C2'));
      expect(res.getTags[1].color, isNull);
    });
  });

  group('list query with enum argument', () {
    test('listUsersByStatus sends enum as string variable', () async {
      adapter.registerData('listUsersByStatus', {'listUsersByStatus': []});
      await client.queries.listUsersByStatus(status: UserStatus.ACTIVE);
      expect(adapter.lastCall!.variables['status'], equals('ACTIVE'));
    });
  });
}

File: integration_tests/typescript_client_tests/test/lists.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { GraphLinkClient } from '../lib/generated/client/graph-link-client.js';
import { UserStatus } from '../lib/generated/enums/user-status.js';
import { MockAdapter, MockWsAdapter } from './mock-adapter.ts';
import { kUserAliceJson, kUserBobJson, kTagsJson } from './fixtures.ts';

let adapter: MockAdapter;
let client: GraphLinkClient;

beforeEach(() => {
  adapter = new MockAdapter();
  client = new GraphLinkClient(adapter.call, new MockWsAdapter());
});

describe('non-nullable list of objects', () => {
  it('listUsers returns a correctly typed list', async () => {
    adapter.registerData('listUsers', { listUsers: [kUserAliceJson, kUserBobJson] });
    const res = await client.queries.listUsers();
    expect(res.listUsers.length).toBe(2);
    expect(res.listUsers[0].id).toBe('user-1');
    expect(res.listUsers[1].id).toBe('user-2');
  });

  it('empty list is returned as empty list, not null', async () => {
    adapter.registerData('listUsers', { listUsers: [] });
    const res = await client.queries.listUsers();
    expect(res.listUsers).not.toBeNull();
    expect(res.listUsers).toHaveLength(0);
  });

  it('each item in the list is fully deserialized', async () => {
    adapter.registerData('listUsers', { listUsers: [kUserAliceJson] });
    const res = await client.queries.listUsers();
    expect(res.listUsers[0].name).toBe('Alice Smith');
    expect(res.listUsers[0].address.city).toBe('Springfield');
  });
});

describe('list of scalars on a type', () => {
  it('User.tags list deserializes correctly', async () => {
    adapter.registerData('getUser', { getUser: kUserAliceJson });
    const res = await client.queries.getUser({ id: 'user-1' });
    expect(res.getUser.tags).toEqual(['admin', 'beta']);
  });

  it('empty tags list is empty, not null', async () => {
    adapter.registerData('getUser', { getUser: kUserBobJson });
    const res = await client.queries.getUser({ id: 'user-2' });
    expect(res.getUser.tags).toHaveLength(0);
  });

  it('nullable scores list is null when server returns null', async () => {
    adapter.registerData('getUser', { getUser: kUserBobJson });
    const res = await client.queries.getUser({ id: 'user-2' });
    expect(res.getUser.scores).toBeNull();
  });

  it('nullable scores list has values when server returns them', async () => {
    adapter.registerData('getUser', { getUser: kUserAliceJson });
    const res = await client.queries.getUser({ id: 'user-1' });
    expect(res.getUser.scores).toEqual([10, 20, 30]);
  });
});

describe('list of objects with nullable fields', () => {
  it('getTags returns tags with nullable color field', async () => {
    adapter.registerData('getTags', { getTags: kTagsJson });
    const res = await client.queries.getTags();
    expect(res.getTags).toHaveLength(2);
    expect(res.getTags[0].label).toBe('dart');
    expect(res.getTags[0].color).toBe('#0175C2');
    expect(res.getTags[1].color).toBeNull();
  });
});

describe('list query with enum argument', () => {
  it('listUsersByStatus sends enum as string variable', async () => {
    adapter.registerData('listUsersByStatus', { listUsersByStatus: [] });
    await client.queries.listUsersByStatus({ status: UserStatus.ACTIVE });
    expect(adapter.lastCall!.variables['status']).toBe('ACTIVE');
  });
});
